---
title: "01-blast-tutorial"
author: "Zach Bengtsson"
date: "5/10/2023"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## 1. Download NCBI Blast

The most current version is 2.14.0. Follow [this link](https://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/) to make sure you are downloading the most recent version.

### 1.1 curl and tar

* `curl` will download the software to your chosen location
* `tar` will unzip your downloaded file so the software may be called and used

The link you grab for your specific operating system goes in the `curl` command. Adjust filename accordingly in your `tar` command to 

```{bash, eval=FALSE}
cd /Applications/bioinfo/
curl -O https://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/ncbi-blast-2.14.0+-x64-macosx.tar.gz
tar -xf ncbi-blast-2.14.0+-x64-macosx.tar.gz
```

### 1.2 Check

Ensure your bioinfo folder contains the unzipped version of NCBI Blast. In this case, ncbi-blast-2.14.0+.

```{bash, eval=FALSE}
ls /Applications/bioinfo/
```

### 1.3 View Usage Information

This code runs the blastx program, which is part of the NCBI Blast program we downloaded. The `-h` option displays the help message for blastx, which provides information about the program's command line options and usage.

```{bash, eval=FALSE}
/Applications/bioinfo/ncbi-blast-2.14.0+/bin/blastx -h
```

## 2. Make a Blast Database

UniProt is a comprehensive, high-quality protein sequence database that provides a central resource for protein sequences and functional information. We will make a blast database using UniProt.

### 2.1 Download Database

Download UniProtKB Swiss-Prot protein sequence database in FASTA format.

* `curl` downloads zipped fasta to location specified by `cd`
* `mv` renames file as uniprot_sprot_r2023_01.fasta.gz
* `gunzip` unzips the file
* `ls` lists the contents of the ../data directory to confirm that the uniprot_sprot_r2023_01.fasta file has been created

```{bash, eval=FALSE}
cd ../data
curl -O https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.fasta.gz
mv uniprot_sprot.fasta.gz uniprot_sprot_r2023_01.fasta.gz
gunzip -k uniprot_sprot_r2023_01.fasta.gz
ls ../data
```

### 2.2 Create Blast Database
Make blast database with the `makeblastdb` program.

* The first line runs the `makeblastdb` program
* `-in` specifies input file is ../data/uniprot_sprot_r2023_01.fasta, which is the UniProt protein database in FASTA format that was downloaded and unzipped in a previous step
* `-dbtype prot` specifies that the database type is protein
* `-out ../blastdb/uniprot_sprot_r2023_01` specifies the name of the output database file and its directory


```{bash, eval=FALSE}
/Applications/bioinfo/ncbi-blast-2.14.0+/bin/makeblastdb \
-in ../data/uniprot_sprot_r2023_01.fasta \
-dbtype prot \
-out ../blastdb/uniprot_sprot_r2023_01
```

## 3. Retrieve Query Sequence

Get the sequence we will be blasting from one of the Roberts Lab's machines.

### 3.1 Download with `curl`

* `-k` allows `curl` to proceed and operate even for server connections considered insecure
* `>` redirects the output to a file named Ab_4denovo_CLC6_a.fa in the ../data/ directory

```{bash, eval=FALSE}
curl https://eagle.fish.washington.edu/cnidarian/Ab_4denovo_CLC6_a.fa \
-k \
> ../data/Ab_4denovo_CLC6_a.fa
```

### 3.2 Check

Inspect the file just downloaded.

* `head` prints the first few lines of the file for viewing
* `echo` prints "How many sequences are there?" to the console
* `grep` searches for lines in the file that start with ">", which is the FASTA file format for sequence headers. By counting the number of matches, it determines the number of sequences in the file and prints that number to the console.

```{bash, eval=TRUE}
head ../data/Ab_4denovo_CLC6_a.fa
echo "How many sequences are there?"
grep -c ">" ../data/Ab_4denovo_CLC6_a.fa
```

## 4. Run Blast

Perform a blastx search of the protein sequences in the input FASTA file (Ab_4denovo_CLC6_a.fa) against the UniProtKB/Swiss-Prot protein database (uniprot_sprot_r2023_01).

### 4.1 Run BlastX

* The first line runs the blastx program
* `-query` specifies the input query FASTA file
* `-db` specifies the reference database against which to search
* `-out` specifies the name and location of the output file for the blastx results
* `-evalue` option sets the maximum e-value threshold for reporting matches
* `-num_threads` sets the number of CPU threads to use for the blast search
* `-max_target_seqs` sets the maximum number of aligned sequences to keep in the output
* `-outfmt` specifies the format of the output (in this case, tabular format)

```{bash, eval=FALSE}
/Applications/bioinfo/ncbi-blast-2.14.0+/bin/blastx \
-query ../data/Ab_4denovo_CLC6_a.fa \
-db ../blastdb/uniprot_sprot_r2023_01 \
-out ../output/Ab_4-uniprot_blastx.tab \
-evalue 1E-20 \
-num_threads 100 \
-max_target_seqs 1 \
-outfmt 6
```

### 4.2 Check

Inspect your output file.

* `head` prints the first few lines for viewing
* `wc -l` counts the number of lines in the file 

```{bash, eval=FALSE}
head -2 ../output/Ab_4-uniprot_blastx.tab
wc -l ../output/Ab_4-uniprot_blastx.tab
```

## 5. Add more Information

### 5.1 

```{bash, eval=FALSE}
curl -o "Accept: text/plain; format=tsv" "https://rest.uniprot.org/uniprotkb/search?query=reviewed:true+AND+organism_id:9606"
```

```{bash, eval=FALSE}
curl -o uniprot_table_r2023_01.tab -H "Accept: text/plain; format=tsv" "https://rest.uniprot.org/uniprotkb/stream?compressed=true&fields=accession%2Creviewed%2Cid%2Cprotein_name%2Cgene_names%2Corganism_name%2Clength%2Cgo_f%2Cgo%2Cgo_p%2Cgo_c%2Cgo_id%2Ccc_interaction%2Cec%2Cxref_reactome%2Cxref_unipathway%2Cxref_interpro&format=tsv&query=%28%2A%29%20AND%20%28reviewed%3Atrue%29"
```

```{bash, eval=FALSE}
head -2 ../output/Ab_4-uniprot_blastx.tab
wc -l ../output/Ab_4-uniprot_blastx.tab
```

```{bash, eval=FALSE}
tr '|' '\t' < ../output/Ab_4-uniprot_blastx.tab | head -2
```

```{bash, eval=FALSE}
tr '|' '\t' < ../output/Ab_4-uniprot_blastx.tab \
> ../output/Ab_4-uniprot_blastx_sep.tab
```

```{bash, eval=FALSE}
head -2 ../code/uniprot_table_r2023_01.tab
wc -l ../code/uniprot_table_r2023_01.tab
```

## 6. Load Packages

```{r, eval=TRUE, echo=FALSE}
library(tidyverse)
library(readr)
library(dplyr)
library(httr)
library(kableExtra)
library(magrittr)
```

## Put Together or Something

```{r, eval=TRUE, echo=FALSE}
bltabl <- read.csv("../output/Ab_4-uniprot_blastx_sep.tab", sep = '\t', header = FALSE)
url <- "https://gannet.fish.washington.edu/seashell/snaps/uniprot_table_r2023_01.tab"
response <- GET(url, config(ssl_verifypeer = FALSE))
spgo <- read.csv(text = rawToChar(response$content), sep = "\t", header = TRUE)
```

```{r, eval=FALSE}
str(spgo)
```

```{r, eval=TRUE, echo=TRUE}
kbl(
head(
  left_join(bltabl, spgo,  by = c("V3" = "Entry")) %>%
  select(V1, V3, V13, Protein.names, Organism, Gene.Ontology..biological.process., Gene.Ontology.IDs) %>% mutate(V1 = str_replace_all(V1, 
            pattern = "solid0078_20110412_FRAG_BC_WHITE_WHITE_F3_QV_SE_trimmed", replacement = "Ab"))
)
) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"))
```

```{r}
left_join(bltabl, spgo,  by = c("V3" = "Entry")) %>%
  select(V1, V3, V13, Protein.names, Organism, Gene.Ontology..biological.process., Gene.Ontology.IDs) %>% mutate(V1 = str_replace_all(V1, 
            pattern = "solid0078_20110412_FRAG_BC_WHITE_WHITE_F3_QV_SE_trimmed", replacement = "Ab")) %>%
  write_delim("../output/blast_annot_go.tab", delim = '\t')
```

```{bash}
head ../output/blast_annot_go.tab
```


## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

```{r cars}
summary(cars)
```

## Including Plots

You can also embed plots, for example:

```{r pressure, echo=FALSE}
plot(pressure)
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
