---
title: "Blast Tutorial"
author: "Zach Bengtsson"
date: "May 14, 2023"
output: 
  html_document:
    theme: readable
    toc: true
    toc_float: true
    number_sections: true
    code_folding: show
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Download NCBI Blast

The most current version is 2.14.0. Follow [this link](https://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/) to make sure you are downloading the most recent version.

## curl and tar

* `curl` will download the software to your chosen location
* `tar` will unzip your downloaded file so the software may be called and used

The link you grab for your specific operating system goes in the `curl` command. Adjust filename accordingly in your `tar` command to 

```{bash, eval=FALSE}
cd /Applications/bioinfo/
curl -O https://ftp.ncbi.nlm.nih.gov/blast/executables/blast+/LATEST/ncbi-blast-2.14.0+-x64-macosx.tar.gz
tar -xf ncbi-blast-2.14.0+-x64-macosx.tar.gz
```

## Check

Ensure your bioinfo folder contains the unzipped version of NCBI Blast. In this case, ncbi-blast-2.14.0+.

```{bash, eval=FALSE}
ls /Applications/bioinfo/
```

## View Usage Information

This code runs the blastx program, which is part of the NCBI Blast program we downloaded. The `-h` option displays the help message for blastx, which provides information about the program's command line options and usage.

```{bash, eval=FALSE}
/Applications/bioinfo/ncbi-blast-2.14.0+/bin/blastx -h
```

# Make a Blast Database

UniProt is a comprehensive, high-quality protein sequence database that provides a central resource for protein sequences and functional information. We will make a blast database using UniProt.

## Download Database

Download UniProtKB Swiss-Prot protein sequence database in FASTA format.

* `curl` downloads zipped fasta to location specified by `cd`
* `mv` renames file as uniprot_sprot_r2023_01.fasta.gz
* `gunzip` unzips the file
* `ls` lists the contents of the ../data directory to confirm that the uniprot_sprot_r2023_01.fasta file has been created

```{bash, eval=FALSE}
cd ../data
curl -O https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/complete/uniprot_sprot.fasta.gz
mv uniprot_sprot.fasta.gz uniprot_sprot_r2023_01.fasta.gz
gunzip -k uniprot_sprot_r2023_01.fasta.gz
ls ../data
```

## Create Blast Database
Make blast database with the `makeblastdb` program.

* The first line runs the `makeblastdb` program
* `-in` specifies input file is ../data/uniprot_sprot_r2023_01.fasta, which is the UniProt protein database in FASTA format that was downloaded and unzipped in a previous step
* `-dbtype prot` specifies that the database type is protein
* `-out ../blastdb/uniprot_sprot_r2023_01` specifies the name of the output database file and its directory


```{bash, eval=FALSE}
/Applications/bioinfo/ncbi-blast-2.14.0+/bin/makeblastdb \
-in ../data/uniprot_sprot_r2023_01.fasta \
-dbtype prot \
-out ../blastdb/uniprot_sprot_r2023_01
```

## Retrieve Query Sequence

Get the sequence we will be blasting from one of the Roberts Lab's machines.

## Download with `curl`

* `-k` allows `curl` to proceed and operate even for server connections considered insecure
* `>` redirects the output to a file named Ab_4denovo_CLC6_a.fa in the ../data/ directory

```{bash, eval=FALSE}
curl https://eagle.fish.washington.edu/cnidarian/Ab_4denovo_CLC6_a.fa \
-k \
> ../data/Ab_4denovo_CLC6_a.fa
```

## Check

Inspect the file just downloaded.

* `head` prints the first few lines of the file for viewing
* `echo` prints "How many sequences are there?" to the console
* `grep` searches for lines in the file that start with ">", which is the FASTA file format for sequence headers. By counting the number of matches, it determines the number of sequences in the file and prints that number to the console.

```{bash, eval=TRUE}
head ../data/Ab_4denovo_CLC6_a.fa
echo "How many sequences are there?"
grep -c ">" ../data/Ab_4denovo_CLC6_a.fa
```

# Run Blast

Perform a blastx search of the protein sequences in the input FASTA file (Ab_4denovo_CLC6_a.fa) against the UniProtKB/Swiss-Prot protein database (uniprot_sprot_r2023_01).

## Run BlastX

* The first line runs the blastx program
* `-query` specifies the input query FASTA file
* `-db` specifies the reference database against which to search
* `-out` specifies the name and location of the output file for the blastx results
* `-evalue` option sets the maximum e-value threshold for reporting matches
* `-num_threads` sets the number of CPU threads to use for the blast search
* `-max_target_seqs` sets the maximum number of aligned sequences to keep in the output
* `-outfmt` specifies the format of the output (in this case, tabular format)

```{bash, eval=FALSE}
/Applications/bioinfo/ncbi-blast-2.14.0+/bin/blastx \
-query ../data/Ab_4denovo_CLC6_a.fa \
-db ../blastdb/uniprot_sprot_r2023_01 \
-out ../output/Ab_4-uniprot_blastx.tab \
-evalue 1E-20 \
-num_threads 100 \
-max_target_seqs 1 \
-outfmt 6
```

## Check

Inspect your output file.

* `head` prints the first few lines for viewing
* `wc -l` counts the number of lines in the file 

```{bash, eval=TRUE}
head -2 ../output/Ab_4-uniprot_blastx.tab
wc -l ../output/Ab_4-uniprot_blastx.tab
```

# Get Annotation Information

## Query the Uniprot Database

To generate the API URL used with `curl` to successfully download the SwissProt protein database for annotating blast results following go to [this Uniprot website link](https://www.uniprot.org/uniprotkb?facets=reviewed%3Atrue&query=%2A). Then click download > change format to TSV > click Generate URL for API > copy the link for API URL using the streaming endpoint. The link has already been placed in the following code chunk. 

The database at this URL contains several parameters that specify the fields to be included in the TSV file, such as accession number, protein name, gene names, organism name, length, and various annotations (e.g., Gene Ontology terms, EC numbers, and cross-references to other databases)

* `curl` downloads a tab-separated values (TSV) file from a URL specified by the given Uniprot query
* `-o` specifies downloaded file is named uniprot_table_r2023_01.tab
* `-H` option is used to specify the "Accept" header of the HTTP request as "text/plain; format=tsv", indicating that the server should return a TSV file

```{bash, eval=FALSE}
curl -o uniprot_table_r2023_01.tab -H "Accept: text/plain; format=tsv" "https://rest.uniprot.org/uniprotkb/stream?compressed=true&fields=accession%2Creviewed%2Cid%2Cprotein_name%2Cgene_names%2Corganism_name%2Clength%2Cgo_f%2Cgo%2Cgo_p%2Cgo_c%2Cgo_id%2Ccc_interaction%2Cec%2Cxref_reactome%2Cxref_unipathway%2Cxref_interpro&format=tsv&query=%28%2A%29%20AND%20%28reviewed%3Atrue%29"
```

# Join Blast and Annotation Tables

## Inspect Blast Table

```{bash, eval=TRUE}
head -2 ../output/Ab_4-uniprot_blastx.tab
wc -l ../output/Ab_4-uniprot_blastx.tab
```

## Process Blast Table

The Blast table needs to be converted to a tab separated table to join with the annotation table, which is already in tab separated format.

* `tr` replaces the vertical bars (|) in the file Blast table (Ab_4-uniprot_blastx.tab) with tabs (\t)
* The resulting output is then redirected to a new file (../output/Ab_4-uniprot_blastx_sep.tab) using the `>` operator
* The output file is then piped to `head -2` to print the first two lines of the file with the replaced separator

```{bash, eval=TRUE}
tr '|' '\t' < ../output/Ab_4-uniprot_blastx.tab \
> ../output/Ab_4-uniprot_blastx_sep.tab

head -2 ../output/Ab_4-uniprot_blastx_sep.tab
```

## Load Packages

These packages are necessary to join and display the two tables.

```{r, eval=TRUE, echo=TRUE, warning=FALSE, message=FALSE}
library(tidyverse)
library(readr)
library(dplyr)
library(httr)
library(kableExtra)
library(magrittr)
```

## Read in Data Tables

The following chunk reads the Blast and annotation tables in as dataframes called `bltabl` and `spgo`. Note that the SwissProt table downloaded from UniProt website is hosted on the Gannet server and read in directly via URL.

```{r, eval=TRUE, echo=TRUE, warning=FALSE, message=FALSE}
bltabl <- read.csv("../output/Ab_4-uniprot_blastx_sep.tab", sep = '\t', header = FALSE)
url <- "https://gannet.fish.washington.edu/seashell/snaps/uniprot_table_r2023_01.tab"
response <- GET(url, config(ssl_verifypeer = FALSE))
spgo <- read.csv(text = rawToChar(response$content), sep = "\t", header = TRUE)
```

## Combine Blast and Annotation Tables

This code creates a table that combines the results of the BlastX and information about the SwissProt entries that were hit by the query by taking the bltabl and spgo data frames and joining them together based on the V3 and Entry columns, respectively. And then displays the results.

* The `select` function is then used to keep only certain columns from the resulting data frame, specifically V1, V3, V13, Protein.names, Organism, Gene.Ontology..biological.process., and Gene.Ontology.IDs
* The `mutate` function is used to rename the V1 column by replacing a specific string of characters within the column names with "Ab"
* The resulting data frame is then passed to the `head` function to return the first few rows, and then to the `kbl` function to format it as a table
* The `kable_styling` function is used to apply Bootstrap styling options to the resulting table

```{r, eval=TRUE, echo=TRUE}
kbl(
head(
  left_join(bltabl, spgo,  by = c("V3" = "Entry")) %>%
  select(V1, V3, V13, Protein.names, Organism, Gene.Ontology..biological.process., Gene.Ontology.IDs) %>% mutate(V1 = str_replace_all(V1, 
            pattern = "solid0078_20110412_FRAG_BC_WHITE_WHITE_F3_QV_SE_trimmed", replacement = "Ab"))
)
) %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"))
```

## Write Table to Tab Separated File

This code writes the joined table to a tab separated file for further use and completes the same functions as the above code.

* Takes the data from the `bltabl` and `spgo` data frames and performs a left join on them, using `V3` in `bltabl` and `Entry` in `spgo` as the joining columns.
* Selects columns `V1`, `V3`, `V13`, `Protein.names`, `Organism`, `Gene.Ontology..biological.process.`, and `Gene.Ontology.IDs` from the joined data frame.
* Replaces all instances of "solid0078_20110412_FRAG_BC_WHITE_WHITE_F3_QV_SE_trimmed" in column `V1` with "Ab" using the `str_replace_all` function.
* Writes the resulting data frame to a tab-delimited file named "blast_annot_go.tab" in the "../output/" directory.

```{r}
left_join(bltabl, spgo,  by = c("V3" = "Entry")) %>%
  select(V1, V3, V13, Protein.names, Organism, Gene.Ontology..biological.process., Gene.Ontology.IDs) %>% mutate(V1 = str_replace_all(V1, 
            pattern = "solid0078_20110412_FRAG_BC_WHITE_WHITE_F3_QV_SE_trimmed", replacement = "Ab")) %>%
  write_delim("../output/blast_annot_go.tab", delim = '\t')
```

# Final Output

The final tab file with annotated blast results can be found [here](https://github.com/course-fish546-2023/zach-coursework/blob/main/assignments/output/blast_annot_go.tab).