---
title: "07-CG-motifs"
author: "Zach Bengtsson"
date: "2023-05-12"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Links to final screenshots:

![image](https://github.com/course-fish546-2023/zach-coursework/blob/main/assignments/output/07-CG-motif-01.png?raw=true)

![image](https://github.com/course-fish546-2023/zach-coursework/blob/main/assignments/output/07-CG-motif-02.png?raw=true)

```{r}
library(seqinr)

# Replace 'input.fasta' with the name of your multi-sequence fasta file
input_file <- "~/github/zach-lncRNA/data/Pver_genome_assembly_v1.0.fasta"
sequences <- read.fasta(input_file)

```


```{r}
# Set the seed for reproducibility (optional)
set.seed(42)

number_of_sequences_to_select <- 10

if (length(sequences) < number_of_sequences_to_select) {
  warning("There are fewer than 10 sequences in the fasta file. All sequences will be selected.")
  number_of_sequences_to_select <- length(sequences)
}

selected_indices <- sample(length(sequences), number_of_sequences_to_select)
selected_sequences <- sequences[selected_indices]

```


```{r}
# Replace 'output.fasta' with your desired output file name
output_file <- "~/github/zach-coursework/output/07-CG-motifs.fasta"
write.fasta(selected_sequences, names(selected_sequences), output_file, open = "w")
```


```{bash}
#likely will not need; fix issue where gff and fa name did not match
# sed -i 's/>lcl|/>/g' ../output/10_seqs.fa
```


```{bash}
#needed downstream for IGV
/home/shared/samtools-1.12/samtools faidx \
~/github/zach-coursework/output/07-CG-motifs.fasta
```


```{bash}
fuzznuc -sequence ~/github/zach-coursework/output/07-CG-motifs.fasta -pattern CG -rformat gff -outfile ~/github/zach-coursework/output/07-CGoutput.gff
```